src/patch.rs

```diff
@@ -0,0 +1,203 @@
+                                                          new_path,
+                                                          opts.map(|s| s.raw())));
+            Ok(Binding::from_raw(ret))
+        }
+    }
+
+    /// Generate a Patch by diffing two buffers.
+    pub fn from_buffers<S1, S2>(old_buffer: &[u8],
+                                old_path: Option<S1>,
+                                new_buffer: &[u8],
+                                new_path: Option<S2>,
+                                opts: Option<&mut DiffOptions>)
+                                -> Result<Patch, Error>
+        where S1: IntoCString, S2: IntoCString
+    {
+        let mut ret = 0 as *mut raw::git_patch;
+        let old_path = try!(into_opt_c_string(old_path));
+        let new_path = try!(into_opt_c_string(new_path));
+        unsafe {
+            try_call!(raw::git_patch_from_buffers(&mut ret,
+                                                 old_buffer.as_ptr() as *const c_char,
+                                                 old_buffer.len(),
+                                                 old_path,
+                                                 new_buffer.as_ptr() as *const c_char,
+                                                 new_buffer.len(),
+                                                 new_path,
+                                                 opts.map(|s| s.raw())));
+            Ok(Binding::from_raw(ret))
+        }
+    }
+
+    /// Get the DiffDelta associated with the Patch.
+    pub fn delta(&self) -> DiffDelta {
+        unsafe {
+            Binding::from_raw(raw::git_patch_get_delta(self.raw) as *mut _)
+        }
+    }
+
+    /// Get the number of hunks in the Patch.
+    pub fn num_hunks(&self) -> usize {
+        unsafe {
+            raw::git_patch_num_hunks(self.raw)
+        }
+    }
+
+    /// Get the number of lines of context, additions, and deletions in the Patch.
+    pub fn line_stats(&self) -> Result<(usize, usize, usize), Error> {
+        let mut context = 0;
+        let mut additions = 0;
+        let mut deletions = 0;
+        unsafe {
+            try_call!(raw::git_patch_line_stats(&mut context,
+                                                &mut additions,
+                                                &mut deletions,
+                                                self.raw));
+        }
+        Ok((context, additions, deletions))
+    }
+
+    /// Get a DiffHunk and its total line count from the Patch.
+    pub fn hunk(&mut self, hunk_idx: usize) -> Result<(DiffHunk, usize), Error> {
+        let mut ret = 0 as *const raw::git_diff_hunk;
+        let mut lines = 0;
+        unsafe {
+            try_call!(raw::git_patch_get_hunk(&mut ret, &mut lines, self.raw, hunk_idx));
+            Ok((Binding::from_raw(ret), lines))
+        }
+    }
+
+    /// Get the number of lines in a hunk.
+    pub fn num_lines_in_hunk(&self, hunk_idx: usize) -> Result<i32, Error> {
```

Comment:
I'd done it this way because the underlying function would break for values more than int.  (It should have taken a `size_t` out parameter and returned an int as an error code only.)  But I'll switch it to `usize`.